pkg/reconciler/pipelinerun/resources/apply.go

```diff
@@ -139,7 +139,8 @@ func paramsFromPipelineRun(ctx context.Context, pr *v1beta1.PipelineRun) (map[st
 	return stringReplacements, arrayReplacements, objectReplacements
 }
 
-func getContextReplacements(pipelineName string, pr *v1beta1.PipelineRun) map[string]string {
+// GetContextReplacements returns the pipelineRun context which can be used to replace context variables in the specifications
+func GetContextReplacements(pipelineName string, pr *v1beta1.PipelineRun) map[string]string {
```

Comment:
now that this function is exported, we should add tests for it

didn't see any tests for it in [apply_test.go](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/resources/apply_test.go), unless I missed something

Reply:
thank you @jerop for the review, we already have a list of tests for validating the context replacements. Please let me know if you prefer these tests to be updated. I think there is no changes needed here but happy to follow up on the review.

https://github.com/tektoncd/pipeline/blob/e8c14bd1ab7075cee2dfcb520cea8a3c8b57a884/pkg/reconciler/pipelinerun/resources/apply_test.go#L2834-L2910

Reply:
Those tests are for `ApplyContexts`, doesn't seem we have tests for `GetContextReplacements`

I think that now we are exporting `GetContextReplacements` and using it independently of `ApplyContexts`, we should add tests for `GetContextReplacements` itself

Reply:
These checks are explicitly testing pipelineRun context variables replacements which is the underlying core functionality of GetContextReplacements.

Reply:
Please refer to the corresponding go coverage report:

![Uploading CB7371ED-38D8-4B16-BD42-3C8A50435F14.jpeg…]()

Reply:
<img width="1323" alt="Screen Shot 2023-03-15 at 9 11 41 PM" src="https://user-images.githubusercontent.com/206285/225512465-5b645619-96ef-46a3-a0bc-21f323e91349.png">

The go coverage [report](https://ffc31bbbcd5c848724809c6645648db8188bb74a1ea5b3e482425cf-apidata.googleusercontent.com/download/storage/v1/b/tekton-prow/o/pr-logs%2Fpull%2Ftektoncd_pipeline%2F6165%2Fpull-tekton-pipeline-go-coverage%2F1625620679015534592%2Fartifacts%2Fline-cov.html?jk=AahUMlvl_wSoAVChaAp48SLpXWYEohjxe0JVceVjvM1gaqk_mXQVxkqCRTxdDGKXBSGNjR_2IP2V6D224Igt7dpC2yWYxw08tW9FCuchQznq67aov8YG79JGzcgWwMrGiddVVFSegLsoC3aQXXUuxOXzyrYe5ObafEZkafEFDz4JoHQvDkYQme5iA7cCTnh3NELwezYcP8Eprp6Xk97MDvfCBwJJzj4zK_atyxUs2YX7O-Ya-8LPuqbmlMJOtVjtMemiY5o_oDndfGOqclnBSyWKFZXgS5DxtJDdrhNB5Mp1VhsqFbcP1d6BoXjmvlA-s3JdVobjW9DK7KL14TdtJ_I16nMsQPBkGkkKwxAMuZ5kdOmtB70UpjPe76v9izLIQHCNOHQm9iz1tdIjEmtJ4NHn9sbLdzaX_k2UDhCsv1qWyFz05L1TY24kUMfZc8l-I0PcSkE6VqW2iIRWv5I3cHS-BUTjdDGo3Mv4l5fUKR4CvRnbQ1le4fMTKj2FPL-yRUT33bdEFQQIMtIYdY9u7sL2zr9o6qLdfxu7foXnpN2wP9_hCbcJrmberLmDHTcX5p4R4kg4Wp6ksp6onuUIeqvmK7BVpvv9J2foBaBwCfC1sPY5heLyHdQ2qgNKu7fRdR-vzCOBUpxwec85vD_kl4y94RK_kyUn9QShK8CENVTY64ROPzOnoW16MPxY3oXJSCR--lsX1nzBc2cZnEtwVV7nu6RRgfLyP2H0Cz1MsgC3cIdReBLu_OKFtJRpG8sKVfOuAMNiNxP-8rUlvixk8y5jTKk0q4Yf2tD2rGZPCkFhOwQwdJYoTvlAv0_KL1gijSfb2YqkBs4YJR2JWjfhOKMZHgi0rqPgMnqX1xJkGH0VWyw6UGF2TmBv2IBElaKvp11s_-r5uaggPKxfdLOLAcQnElkqa1pgua7ObkfaZTJmSPRSryj6xrR9-zCysQGiCUGYy4xMmeYMxlvxhID_43p1dlFXsjduwALM2FQ0XfyJLBmlMElVi82_YSN11u9zCKXe6_5s7D3zChL4IJUmzRXw53Tw80ueMtHArzGTOXxi3prW0fiMU4lQoDhNw36jeRK9oioPXtuO6VO82zkEfL7sLwcY7O4Pq4TPfWD-JowoSEAhHHpUmDCv-MVX7jR6sIRzE3hWH_-I3ZSh8JE7YuVtGv2mb7ikohTNH8Y58y2BSUm8O0In2Zb9__4P8QKS2RjkAS0_kwTq-rOJMmL9PgP5y18S5xMgr4eMrMAZ220Nlj9Z9k_soVNBa8swJG-6noHeUnMsYY87x6oN2zkJagJVQimc2VGYjTJQ4yEotaK-DczyI6m4hG68iJ7mWNddI1o-dbzqMq0R1686_szMAQ&isca=1#file0) of this PR.

Reply:
hoping to have this merged for this release 🙏

Reply:
Thanks @jerop for this careful and thorough review. Wish such reviews and reviewers existed across the board and were applied to every PR in pipeline repo.

In the interest of the feature’s availability in the upcoming release and the amount of time spent on the review, I will open an issue to check and fix all existing exported functions having unit tests. 🙏

Reply:
Hi @pritidesai, how about adding a ut like this one? If you're ok with it maybe just copy paste into the pr or with some changes. Or we could leave it to an issue to fix. I'm ok with either. 

This is a small function and covered by another function so this comment may not be a block. But I think @jerop makes a very good point that we should test all exported functions.
```

func TestGetContextReplacements(t *testing.T) {
	for _, tc := range []struct {
		description  string
		pr           *v1beta1.PipelineRun
		pipelineName string
		expected     map[string]string
	}{{
		description:  "pipelineName defined",
		pr:           &v1beta1.PipelineRun{},
		pipelineName: "pipelineName",
		expected:     map[string]string{"context.pipeline.name": "pipelineName"},
	}, {
		description: "pipelineRun defined",
		pr: &v1beta1.PipelineRun{
			ObjectMeta: metav1.ObjectMeta{Name: "name", Namespace: "namespace", UID: "UID"},
		},
		expected: map[string]string{"context.pipelineRun.name": "name", "context.pipelineRun.namespace": "namespace", "context.pipelineRun.uid": "UID"},
	}, {
		description: "both undefined",
		pr:          &v1beta1.PipelineRun{},
		expected:    map[string]string{},
	}} {
		t.Run(tc.description, func(t *testing.T) {
			got := GetContextReplacements(tc.pipelineName, tc.pr)
			keys := []string{"context.pipelineRun.name", "context.pipeline.name", "context.pipelineRun.namespace", "context.pipelineRun.uid"}
			for _, k := range keys {
				if d := cmp.Diff(tc.expected[k], got[k]); d != "" {
					t.Errorf("got different value for %s:%s", k, diff.PrintWantGot(d))
				}
			}
		})
	}
}
```

Reply:
What is the difference between unit tests and code coverage?

Reply:
thank you @Yongxuanzhang 🙇‍♀️ appreciate you jumping in and going beyond to write a test case ⬆️ Are we open to merge this PR so that it can be part of the release?